# Dawson Pallotta

# Update filename and headers

Answer questions in markdown cells. Please verify that your answers are formatted such that they are easy to read.

In [1]:
#import packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.decomposition import PCA
from sklearn.utils import resample
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.metrics import confusion_matrix
from sklearn import tree
import xgboost as xgb
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeRegressor
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.metrics import mean_squared_error, r2_score

In [2]:
#import and read data csv
df1 = pd.read_csv("C:/Users/student/Data Minning/featuresDefault.csv") 
print(df1.shape)
df1.head(2)

(114, 199)


id  phq9  NumTexts  help  office  dance  money  wedding  domestic_work  \
0   e122     7         2   0.0     0.0    0.0    0.0      0.0            0.0   
1  e1526    20         8   0.0     0.0    0.0    0.0      0.0            0.0   

   sleep  ...  monster  ocean  giving  contentment  writing  rural  \
0    0.0  ...      0.0    0.0     0.0          0.0      0.0    0.0   
1    0.0  ...      0.0    0.0     0.0          0.0      0.0    0.0   

   positive_emotion  musical  colors  text_abbreviations  
0               0.0      0.0     0.0                 0.0  
1               0.0      0.0     0.0                 0.0  

[2 rows x 199 columns]

In [3]:
#creates models using cross-validation and scored with MAE
folds = 3

reg = LinearRegression()
lasso = Lasso()
ridge = Ridge() 

scoresreg = cross_val_score(reg, df1[df1.columns[3:]],  df1[df1.columns[1]], cv=folds, scoring = 'neg_mean_absolute_error')
scoreslasso = cross_val_score(lasso, df1[df1.columns[3:]],  df1[df1.columns[1]], cv=folds, scoring = 'neg_mean_absolute_error')
scoresridge = cross_val_score(ridge, df1[df1.columns[3:]],  df1[df1.columns[1]], cv=folds, scoring = 'neg_mean_absolute_error')

print(scoresreg)
print(scoresreg.mean())
print(scoreslasso)
print(scoreslasso.mean())
print(scoresridge)
print(scoresridge.mean())

[-1.71506987e+13 -1.32360755e+13 -1.00420182e+13]
-13476264134485.072
[-6.57894737 -5.75069252 -4.86842105]
-5.732686980609418
[-6.57944857 -5.72338479 -4.86099045]
-5.721274602082023


In [4]:
#creates models using cross-validation and scored with RMSE
folds = 3

reg = LinearRegression()
lasso = Lasso()
ridge = Ridge() 

scoresreg = cross_val_score(reg, df1[df1.columns[3:]],  df1[df1.columns[1]], cv=folds, scoring = 'neg_root_mean_squared_error')
scoreslasso = cross_val_score(lasso, df1[df1.columns[3:]],  df1[df1.columns[1]], cv=folds, scoring = 'neg_root_mean_squared_error')
scoresridge = cross_val_score(ridge, df1[df1.columns[3:]],  df1[df1.columns[1]], cv=folds, scoring = 'neg_root_mean_squared_error')

print(scoresreg)
print(scoresreg.mean())
print(scoreslasso)
print(scoreslasso.mean())
print(scoresridge)
print(scoresridge.mean())

[-3.15043289e+13 -2.13434276e+13 -1.41237588e+13]
-22323838406012.73
[-7.55067139 -6.69405874 -6.00219258]
-6.74897423927591
[-7.54469473 -6.66927014 -5.99099872]
-6.734987862439582


## Question Block 1

These questions regard the above code.
* What is the name of the train/test split strategy being used? How much of the data is in each fold?
cross-validation, 1/3 of the data in each fold - 38
* What are the two metrics being compared?
MAE and RMSE
* Based on the metrics, does the data have many outliers? Explain your reasoning. 
Based on the metrics the data doesn't seem to have a lot of outliers, because of the way the ridge regression works if there were many outliers there would be a big difference between ridge and lasso score but because they are very close in their score, it shows that their isn't many outliers in the dataset
* What do the very high scores from linear regression indicate? Especially in relation to lasso and ridge regression? In other words, why are the scores for linear regression so high?
The model for the linear regression seems to be overfit. It appears that the number of features are larger than the  number of incatnces that are being used meaning that the ability to interpret the coefficients is being lowered by the models multicolinearity
* Is lasso or ridge regression more effective for this problem? Explain your reasoning. 
Risge has a lower MAE than lasso, which makes it a more effective model
* Explain the interpretation of MAE results for lasso and ridge regression in this domain. (ie, the PHQ-9 score is...) 
The lasso and ridge regression models are being used to guess the PHQ-9 scores, they have an error of 6 or 7 points, leading them to predict the scores within 7 points

In [5]:
#linear regression using holdout
df_train, df_test = train_test_split(df1, test_size=0.33, random_state = 33)
reg.fit(df_train[df_train.columns[3:]], df_train[df_train.columns[1]])
regpred = reg.predict(df_test[df_test.columns[3:]])
print("Coefficients: \n", reg.coef_)
print("Mean squared error: %.2f" % mean_squared_error(df_test[df_test.columns[1]], regpred))
print("Coefficient of determination: %.2f" % r2_score(df_test[df_test.columns[1]], regpred))

Coefficients: 
 [-1.19706811e+14 -3.56792268e+14  2.12644719e+15 -1.23495501e+15
 -9.75243164e+14 -1.49781173e+15 -1.56447580e+14 -6.55035894e+15
 -3.73701975e+14 -6.72720336e+14 -9.06137310e+14 -2.63988229e+14
  1.17004202e+14 -9.06257736e+14 -1.09743276e+14  4.96658656e+14
 -1.85411124e+13  1.92904347e+15 -1.04151067e+15 -4.12951448e+14
  1.34219331e+14  1.57968195e+15  8.59185794e+14 -4.11894755e+14
  3.83879405e+14  1.39163100e+14 -3.66285375e+14  4.26247086e+14
  2.53747528e+14 -9.92893704e+13 -9.70217581e+13  2.69716750e+14
 -1.71965289e+14 -2.71471310e+14  2.94710620e+14 -2.33699269e+14
 -1.74430824e+14 -8.36341960e+13 -4.50328372e+14  2.67627922e+14
  9.05703538e+14 -5.10195012e+14 -1.87156635e+14  3.07140524e+14
 -1.56280120e+14 -1.62154021e+14  4.99300371e+14 -1.34579108e+15
 -3.02270141e+14 -1.25197456e+14  5.86405053e+14  2.75849922e+12
 -3.32688365e+14  1.69401205e+14 -9.10029204e+12 -2.10459659e+14
 -5.39785085e+14 -7.67984853e+13 -2.63811281e+13 -9.28373617e+13
  2.96631

In [6]:
#holdout using lasso regression
df_train, df_test = train_test_split(df1, test_size=0.33, random_state = 33)
lasso.fit(df_train[df_train.columns[3:]], df_train[df_train.columns[1]])
lassopred = lasso.predict(df_test[df_test.columns[3:]])
print("Coefficients: \n", lasso.coef_)
print("Mean squared error: %.2f" % mean_squared_error(df_test[df_test.columns[1]], lassopred))
print("Coefficient of determination: %.2f" % r2_score(df_test[df_test.columns[1]], lassopred))

Coefficients: 
 [-0.  0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0. -0. -0. -0.
  0.  0. -0.  0.  0. -0. -0. -0. -0.  0. -0. -0.  0. -0. -0. -0. -0. -0.
 -0. -0.  0. -0. -0.  0.  0. -0. -0. -0. -0.  0. -0. -0.  0. -0. -0. -0.
 -0.  0.  0. -0.  0.  0. -0. -0.  0. -0. -0. -0. -0.  0. -0.  0. -0. -0.
 -0.  0.  0.  0. -0. -0. -0. -0.  0. -0. -0. -0. -0.  0. -0. -0. -0.  0.
 -0.  0. -0.  0.  0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0. -0.
 -0. -0.  0.  0. -0.  0. -0. -0.  0. -0. -0. -0.  0. -0. -0. -0. -0. -0.
  0. -0. -0.  0. -0.  0. -0.  0. -0. -0. -0. -0.  0. -0. -0. -0.  0. -0.
  0. -0. -0. -0.  0. -0. -0. -0. -0. -0.  0. -0. -0. -0.  0. -0. -0.  0.
 -0. -0.  0. -0.  0.  0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0. -0.
 -0.  0.  0. -0. -0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0. -0.]
Mean squared error: 45.57
Coefficient of determination: -0.07


In [7]:
#holdout using ridge regression
df_train, df_test = train_test_split(df1, test_size=0.33, random_state = 33)
ridge.fit(df_train[df_train.columns[3:]], df_train[df_train.columns[1]])
ridgepred = ridge.predict(df_test[df_test.columns[3:]])
print("Coefficients: \n", ridge.coef_)
print("Mean squared error: %.2f" % mean_squared_error(df_test[df_test.columns[1]], ridgepred))
print("Coefficient of determination: %.2f" % r2_score(df_test[df_test.columns[1]], ridgepred))

Coefficients: 
 [-4.62043981e-01  1.32566123e+00 -1.88702718e-01 -1.30681884e+00
 -1.00681847e+00 -9.04190811e-01 -5.31648254e-01 -1.56453495e-01
 -2.15660470e-01 -3.52098017e-01 -2.31721007e-03 -7.24903410e-02
  1.28067528e+00  5.52292662e-03  2.63193868e-01 -3.91118708e-01
 -1.40995974e+00 -1.02586002e-02  1.70823933e-01  1.45376643e-01
 -8.08876407e-02  3.81767780e-02  8.97345550e-03 -1.76614931e-01
 -3.70413207e-01 -1.06239920e-01 -8.58102172e-02  2.87739835e-01
 -2.06869533e-01 -7.24876315e-01  4.82276288e-01 -3.08991347e-01
 -2.33478571e-01 -8.13187621e-02 -3.61884350e-01 -1.60257237e-01
 -9.64887136e-03 -7.02524651e-02  1.76170389e-02 -9.42683082e-01
 -1.35404280e+00  1.21591564e-01  1.10439582e-01 -5.54765849e-01
 -8.54029072e-02 -1.11945479e-01 -2.12247089e-01  1.92966020e-01
 -2.16738765e-01 -2.41877969e-01  3.54858336e-02 -3.20806214e-01
 -6.32687818e-02 -4.76747610e-01 -2.60063720e+00  2.48768046e-01
  2.19247261e-02 -1.72546894e-01  3.93682048e-04  2.50461782e-03
 -3.78304

In [8]:
#displays coefficients (factor variability) for each category, and displays highest and lowest coefficients
#shows the coefficents for each feateure, and displays the highest and lowest coefficient
rc = ridge.coef_
print("This is the maximum coefficient: " + str(max(abs(rc))))
print("This is the minimum coefficient: " + str(min(abs(rc))))
for i in range(0, len(rc)):
    print(rc[i])
    print(df1.columns[i+3])

This is the maximum coefficient: 4.524625570293396
This is the minimum coefficient: 0.0002152105461658456
-0.46204398132712665
help
1.3256612299068617
office
-0.18870271828982704
dance
-1.3068188379144332
money
-1.0068184671485065
wedding
-0.9041908113036086
domestic_work
-0.5316482541808403
sleep
-0.15645349520291102
medical_emergency
-0.21566047014138673
cold
-0.35209801693921194
hate
-0.0023172100707449794
cheerfulness
-0.0724903409846211
aggression
1.2806752799028358
occupation
0.005522926623206627
envy
0.2631938677941391
anticipation
-0.39111870801421533
family
-1.4099597354774975
vacation
-0.010258600168336629
crime
0.17082393321505618
attractive
0.14537664317797344
masculine
-0.08088764067665305
prison
0.038176777951713536
health
0.00897345550411039
pride
-0.17661493146676188
dispute
-0.37041320723911964
nervousness
-0.10623992021529838
government
-0.08581021717887863
weakness
0.28773983472556475
horror
-0.20686953268124175
swearing_terms
-0.724876315474466
leisure
0.48227628793

## Question Block 2

These questions regard the above code.
* What is the name of the train/test split strategy being used? How much data is in the test set?
Holdout, 1/3 of the data os being held in the test set
* Compare the coefficients between the three types of regression. Is this what is expected? Explain.
Lasso: -.07
Ridge: -.07
Regression: -53092483331783856415047680
Yes, the relationship between the two aspects pof lasso and ridge make sense as their variability is similar
* Based on the mean squared error and coefficient of determination, which regularization is best for this problem?
Based on lasso and ridge's similarity of their coefficient of determination, shows that the coefficient of determination is better for this problem than the linear regression
* What does the 0 coefficients for the lasso regression indicate about the features?
It indicates that the features are not included in the  model
* Which feature has the most influence on the prediction based on the ridge regression coefficients? (hint: you can use [control][f])
Communication has the highest coefficient in the dataset with a coeffiecent of 4.524625570293396, meaning that it has the most influence ofver prediction
* What has this lab taught you about regression and regularization?
This lab has taught me that it is incredibly imoortant to test your models multiple times and against other models in order to find out if the model is effecient and accurate while using regression, and using regularlization helps to generalize your model to make better predictions